File: learnLeetCode_remote/Day16/Part1/min_cost_split.cpp
#include <bits/stdc++.h>
using namespace std;

// LC 2547 — Minimum Cost to Split Array
// O(n^2) DP reference implementation

class Solution {
public:
    int minCost(vector<int>& nums, int k) {
        int n = (int)nums.size();
        const long long INF = (long long)4e18;
        vector<long long> dp(n+1, INF);
        dp[0] = 0;
        for (int i = 1; i <= n; ++i) {
            unordered_map<int,int> freq;
            long long dup = 0;
            for (int j = i - 1; j >= 0; --j) {
                int x = nums[j];
                int f = ++freq[x];
                if (f == 2) dup += 2;
                else if (f > 2) dup += 1;
                dp[i] = min(dp[i], dp[j] + k + dup);
            }
        }
        return (int)dp[n];
    }
};

int main(){
    vector<int> nums = {1,2,1,2,1,3,3};
    cout << Solution().minCost(nums, 2) << "\n"; // 8
    return 0;
}

File: learnLeetCode_remote/Day16/Part1/min_cost_split.py
from typing import List

# LC 2547 — Minimum Cost to Split Array
# Cost of subarray = k + number_of_duplicate_elements_in_subarray
# Duplicates counting rule: for an element with frequency f
#   contributes f if f >= 2 else 0
# Incremental: when a frequency hits 2 -> +2; when >2 -> +1
# DP: dp[i] = min cost to split nums[:i]
#   dp[0] = 0
#   dp[i] = min_{j<i} dp[j] + k + duplicates(nums[j:i])
# Note: This O(n^2) reference solution is straightforward and clear.

class Solution:
    def minCost(self, nums: List[int], k: int) -> int:
        n = len(nums)
        dp = [10**18] * (n + 1)
        dp[0] = 0
        for i in range(1, n + 1):
            freq = {}
            dup = 0
            # build backwards segment [j, i)
            for j in range(i - 1, -1, -1):
                x = nums[j]
                f = freq.get(x, 0) + 1
                freq[x] = f
                if f == 2:
                    dup += 2
                elif f > 2:
                    dup += 1
                dp[i] = min(dp[i], dp[j] + k + dup)
        return dp[n]

if __name__ == "__main__":
    print(Solution().minCost([1,2,1,2,1,3,3], 2))  # Expected 8
